checkbox-support/checkbox_support/tests/test_lxd_support.py

```diff
@@ -0,0 +1,337 @@
+                self_mock, "https://ubuntu.com/image", "/tmp/image"
+            )
+        self.assertEqual(isfile_mock.call_count, 1)
+        self.assertTrue(urlretrieve_mock.called)
+
+    @patch("os.path.isfile", return_value=False)
+    @patch("urllib.request.urlretrieve")
+    def test_download_image_file_not_found(
+        self, urlretrieve_mock, isfile_mock, logging_mock
+    ):
+        self_mock = MagicMock()
+        with self.assertRaises(FileNotFoundError):
+            LXD.download_image(
+                self_mock, "https://ubuntu.com/image", "/tmp/image"
+            )
+        self.assertEqual(isfile_mock.call_count, 2)
+        self.assertTrue(urlretrieve_mock.called)
+
+    def test_insert_images_local_success(self, logging_mock):
+        self_mock = MagicMock(template="/tmp/template", image="/tmp/image")
+        LXD.insert_images(self_mock)
+
+    @patch("checkbox_support.lxd_support.run_with_retry")
+    def test_insert_images_remote_success(
+        self, logging_mock, run_with_retry_mock
+    ):
+        self_mock = MagicMock(template=None, image=None, remote="ubuntu:")
+        try:
+            LXD.insert_images(self_mock)
+        except RuntimeError:
+            self.fail("insert_images raised RuntimeError")
+
+    @patch(
+        "checkbox_support.lxd_support.run_with_retry",
+        side_effect=subprocess.CalledProcessError(1, "", ""),
+    )
+    def test_insert_images_remote_fail(
+        self, logging_mock, fake_run_with_retry_mock
+    ):
+        self_mock = MagicMock(template=None, image=None, remote="ubuntu:")
+        with self.assertRaises(subprocess.CalledProcessError):
+            LXD.insert_images(self_mock)
+
+    def test_init_lxd_already_running(self, logging_mock):
+        self_mock = MagicMock()
+        self_mock.run.return_value = MagicMock(returncode=0)
+        LXD.init_lxd(self_mock)
+        self.assertEqual(self_mock.run.call_count, 1)
+
+    def test_init_lxd_success(self, logging_mock):
+        self_mock = MagicMock()
+        self_mock.run.side_effect = [
+            subprocess.CalledProcessError(1, "", "lxd not running"),
+            MagicMock(returncode=0),
+        ]
+        LXD.init_lxd(self_mock)
+        self.assertEqual(self_mock.run.call_count, 2)
+        self.assertTrue(self_mock.insert_images.called)
+
+    def test_cleanup_success(self, logging_mock):
+        self_mock = MagicMock()
+        LXD.cleanup(self_mock)
+
+    def test_launch_no_options(self, logging_mock):
+        self_mock = MagicMock()
+        self_mock.name = "testbed"
+        self_mock.image_alias = MagicMock(
+            hex="656382d4-d820-4d01-944b-82b5b63041a7"
+        )
+        LXD.launch(self_mock)
+
+    def test_launch_options(self, logging_mock):
+        self_mock = MagicMock()
+        self_mock.name = "testbed"
+        self_mock.image_alias = MagicMock(
+            hex="656382d4-d820-4d01-944b-82b5b63041a7"
+        )
+        LXD.launch(self_mock, ["-d root,size=50GB"])
```

Comment:
I would check if the run command is called with the expected arguments here.
One could argue that for this simple function is a bit redundant to check almost the same command in the test as the one you wrote on the code. But I think for this one, which concatenates some options, it makes sense to compare against the expected command. Also it would help you to detect errors in the future if you make this function more complex.
```suggestion
    def test_launch_options(self, logging_mock):
        self_mock = MagicMock()
        self_mock.name = "testbed"
        self_mock.image_alias = MagicMock(
            hex="656382d4-d820-4d01-944b-82b5b63041a7"
        )
        LXD.launch(self_mock, ["-d root,size=50GB"])
        self_mock.run.assert_called_with(
            "lxc launch 656382d4-d820-4d01-944b-82b5b63041a7 testbed -d root,size=50GB"
            )
```
For the following ones, I leave it up to you.
If the lxd command changes , it is double the work to change them. But on the other hand, it could catch a typo if we modify these functions in the future.

Reply:
I've added it for two of the tests in which we add extra options.